### PR TITLE
Resolve #2 Implement `usePack`

### LINKED ISSUE #2: Add a syntax for "packed" structs
Need to think of a good API for it first...

### PR DESCRIPTION
**Pull request by d3m3vilurr**
- Apply [Alexander Seliverstov's method][1]
- Now, `Struct` use 2nd arguments
- `usePack()` switch padding mode

[1]: https://groups.google.com/d/msg/nodejs/2wKCYmGPT5E/DLi4Pjt7j_4J

### CODE REVIEW

**Comment by d3m3vilurr** [comment]
@TooTallNate can you review this pr?

File: test/struct.js [hunk 1]
```diff
@@ -379,4 +379,29 @@ describe('Struct', function () {
 
   })
 
+  describe('usePack()', function () {
+    var size = {
+        unpack: { ia32: 8, x64: 16 },
```
Comment (TooTallNate) [inline test/struct.js]: There's also `arm`, and possible others. Maybe we should do an `is64Bit` check by checking `ref.sizeof.pointer === 8` and then fall back to 32-bit if not.

Comment (d3m3vilurr) [inline test/struct.js]: hmm... basically `padding` depends on compiler. so i think, we should define struct in c file, and compare this.
`ref.sizeof.pointer` is only can check pack value.

this test case should change to
```javascript
{
  unpack: bindings.sizeof_with_padding_struct,
  pack: ref.sizeof.char + ref.sizeof.pointer // or bindings.sizeof_without_padding_struct
}
```
right?

Comment (TooTallNate) [inline test/struct.js]: Ya I think you're right, we should go the native route. Add the struct definition, packed, to the C++ test file, and then export the sizeof and alignof to compare when packed.

File: lib/struct.js [hunk 2]
```diff
@@ -184,6 +186,16 @@ function toString () {
 }
 
 /**
+ * Change struct padding mode.
+ *
+ * @api public
+ */
+function usePack (flag) {
```
Comment (TooTallNate) [inline lib/struct.js]: I don't see much value in having this be an instance method. Why would someone need to change the packed-ness during runtime? C doesn't support this so I'm not sure we need to.

File: test/struct.js (hunk 1 above)
Comment (d3m3vilurr) [inline test/struct.js]: ok, i will try to change tomorrow :)

File: lib/struct.js (hunk 2 above)
Comment (d3m3vilurr) [inline lib/struct.js]: yeah.... actually that is not need.

this is design problem, i think, `Struct`'s  second param have low readability
i like python way like this; `struct = Struct(spec, packed=True)`
btw, current javascript not support named param yet,

so sometimes i will use this instead that
```javascript
var struct = new Struct(spec);
struct.usePack(true);
```
but also, i'm not to change on runtime env

2nd idea, introduce new class like `PackedStruct` is more than good solution about this problem.

few month ago, i was only had first idea... :(

File: lib/struct.js [hunk 3]
```diff
@@ -121,8 +121,10 @@ function Struct () {
   StructType.size = 0
   StructType.alignment = 0
   StructType.indirection = 1
+  StructType.isPacked = arguments[1] || false
```
Comment (TooTallNate) [inline lib/struct.js]: Out of bounds `arguments` checks de-opt the V8 compiler. Do something like this instead:

``` js
arguments.length > 1 ? Boolean(arguments[1] : false
```

**Comment by TooTallNate** [comment]
Thanks for the pull request, and sorry for taking so long to review! This is looking pretty nice so I'll be sure to merge once it's ready.

File: lib/struct.js (hunk 2 above)
Comment (d3m3vilurr) [inline lib/struct.js]: @TooTallNate I will try to second way, but this library direct export `Struct`
So, I will use `Struct.Packed` instead `PackedStruct`, and remove `usePack`, `StructType.isPacked` assign using arguments,  is this ok?
If not, i will just remove `usePack` method

Comment (TooTallNate) [inline lib/struct.js]: Let's just make the 2nd argument be an options Object then, with `packed: true` being one of the options.

``` js
var struct = new Struct(spec, { packed: true });
```

Comment (d3m3vilurr) [inline lib/struct.js]: Ok, I like it :)

**Comment by d3m3vilurr** [comment]
it look appveyor's problem..

**Comment by TooTallNate** [comment]
Thanks @d3m3vilurr! Merged in eb4550fdaa38a0103c1c22fd8490e1204a56685c.

**Pull request closed without merging**